idl/raft_storage.idl.hh

```diff
@@ -22,6 +21,18 @@ struct tagged_id {
 
 } // namespace internal
 
+struct term_t {
```

Comment:
Well, this is not what I thought you were proposing and do not see how it addresses my concerns at all.

I thought you will define `tagged_uint64` as:


```
template<typename Tag>
struct tagged_uint64 : public utils::tagged_integer<Tag, uint64_t> {};
```

Reply:
You replied "let's do it" to "We explicitly defined types for the types that use it", so I went with this direction.

About inheritance, I did indeed proposed that in an earlier comment, but came to the conclusion it's the same as using `using`. The compiler will implicitly defer to resolving `utils::tagged_integer` in the absence of a serializer overload for `tagged_uint64`.

Reply:
> Why does it matter to your concern?

Because if a compiler thinks that those are two different types it will never try to call a wrong function unless one is explicitly convertible to another.

In your example if you comment out:

```
template <typename Tag>
void bar(const tagged_int<Tag>& v) {
    fmt::print("bar tagged_int: {}\n", v.value);
}
```

It will happily choose another (wrong) overload.

Reply:
> > Why does it matter to your concern?
> 
> Because if a compiler thinks that those are two different types it will never try to call a wrong function unless one is explicitly convertible to another.
> 
> In your example if you comment out:
> 
> ```
> template <typename Tag>
> void bar(const tagged_int<Tag>& v) {
>     fmt::print("bar tagged_int: {}\n", v.value);
> }
> ```
> 
> It will happily choose another (wrong) overload.

Right, and it happens in both cases: `using tagged_uint64 = utils::tagged_integer` or with
`struct tagged_uint64 : public utils::tagged_integer`.
Therefore it doesn't help with anything to use inheritance.

Reply:
> 
> Right, and it happens in both cases: `using tagged_uint64 = utils::tagged_integer` or with `struct tagged_uint64 : public utils::tagged_integer`. Therefore it doesn't help with anything to use inheritance.

So how can we make `tagged_uint64` not convertible to `utils::tagged_integer`? Because if it is with inheritance you are right that it is not better.

Reply:
Yeah, this should work.

Reply:
done in v3